src/charm.py

```diff
@@ -35,85 +33,194 @@ def __init__(self, *args: typing.Any):
+        flush table ip aproxy
+        table ip aproxy {{
+            chain prerouting {{
+                type nat hook prerouting priority dstnat; policy accept;
+                {no_proxy_clause}
+                ip daddr != {{ 127.0.0.0/8, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16 }} \
+                    tcp dport {{ {ports_clause} }} counter dnat to 127.0.0.1:8443
+            }}
+
+            chain output {{
+                type nat hook output priority -150; policy accept;
+                ip daddr != {{ 127.0.0.0/8, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16 }} \
+                    tcp dport {{ {ports_clause} }} counter dnat to 127.0.0.1:8443
+            }}
+
+            chain input {{
+                type filter hook input priority filter; policy accept;
+                tcp dport 8443 drop
+            }}
+        }}
         """
-        # Fetch the new config value
-        log_level = str(self.model.config["log-level"]).lower()
-
-        # Do some validation of the configuration option
-        if log_level in VALID_LOG_LEVELS:
-            # The config is good, so update the configuration of the workload
-            container = self.unit.get_container("httpbin")
-            # Verify that we can connect to the Pebble API in the workload container
-            if container.can_connect():
-                # Push an updated layer with the new config
-                container.add_layer("httpbin", self._pebble_layer, combine=True)
-                container.replan()
-
-                logger.debug("Log level for gunicorn changed to '%s'", log_level)
-                self.unit.status = ops.ActiveStatus()
-            else:
-                # We were unable to connect to the Pebble API, so we defer this event
-                event.defer()
-                self.unit.status = ops.WaitingStatus("waiting for Pebble API")
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = ops.BlockedStatus("invalid log level: '{log_level}'")
-
-    @property
-    def _pebble_layer(self) -> pebble.LayerDict:
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
-            "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {
-                        "GUNICORN_CMD_ARGS": f"--log-level {self.model.config['log-level']}"
-                    },
-                }
-            },
-        }
+        try:
+            # nosec B404,B603,B607: calling trusted system binary with predefined args
+            subprocess.run(["nft", "-f", "-"], input=rules.encode(), check=True)  # nosec
```

Comment:
This is not persistent, and the nftables rules will be removed on server reboot. You can consider creating a systemd service to apply the nftables rules.